src/lib/adapters/waku/wakustore.ts

```diff
@@ -115,23 +115,24 @@ export function makeWakustore(waku: LightNode): Wakustore {
 
 	async function onSnapshot<T>(
 		query: Query,
-		callback: (value: T, decodedMessage: DecodedMessage) => void,
+		callback: (value: T, decodedMessage: DecodedMessage) => void | Promise<void>,
 	): Promise<Unsubscribe> {
-		const subscription = await subscribe(waku, query.decoder, (msg: DecodedMessage) => {
+		const subscription = await subscribe(waku, query.decoder, async (msg: DecodedMessage) => {
 			const typedResult = decodeDoc<T>(msg)
-			callback(typedResult, msg)
+			await callback(typedResult, msg)
 		})
 
 		const queryOptions = {
 			...query.queryOptions,
 			pageSize: query.queryOptions?.pageSize ?? query.queryOptions?.limit,
 		}
+
 		const result = await readStore(waku, query.decoder, queryOptions)
 		const messages = await getQueryResults(result, queryOptions)
 
 		for (const message of messages) {
 			const value = decodeDoc<T>(message)
-			callback(value, message)
+			await callback(value, message)
```

Comment:
I wonder if we need to have some validation here. What if the message is not of type `T`?

Reply:
Currently there is no validation at all, so there can be two different types of error:
1. If the message payload is not JSON then `decodeDoc` will fail with an error.
2. If the message is JSON but the type is not `T` then most likely the `callback` will fail.

It would be nice to have validation here and because this is a generic function the solution could be to pass in a generic (optional?) validator function that fails early.

Reply:
Yeah I did that approach in the read from localstore helper fuction. It takes optional schema which it validates against.